### PR TITLE
More tests

### LINKED ISSUE #70: "run_not_found" test is failing
```
___ test_run_not_found ___

    def test_run_not_found():
        status, result, again = run(['notfound'])
        assert status == 'not found'
>       assert result == "[Errno 2] No such file or directory: 'notfound'"
E       assert "[Errno 2] No...': 'notfound'" == "[Errno 2] No ...y: 'notfound'"
E         Skipping 36 identical leading characters in diff, use -v to show
E         -  'notfound': 'notfound'
E         +  'notfound'
```

`-` is actual output, `+` is expected.

So, it's getting `[Errno 2] No such file or directory: 'notfound': 'notfound'` instead of `[Errno 2] No such file or directory: 'notfound'`.

I've disabled the test in 234ef668b74cc20f5f302c6b950e36d26d68bdd5

### PR DESCRIPTION
**Pull request by maxnz**
- **Fix bug where test results are not added to output file**
- Test coverage to 100%
- Remove bidict dependency
- Reincorporate html formatter
- Update gist functionality
- Fixes #70 

More detailed:
- Rename `check_submit_date.py` to `get_modification_time.py`
- Don't record "not a git repository" as an unmerged branch
- Add `FormatType` to represent if Markdown or HTML is used
- Add `FormattedResult` to represent a formatted assignment
- Add `GroupType` to represent how the files should be grouped
- Update html formatter
  - Add free template to make it look nice
  - Add anchors to make jumping to a specific student possible
- Create `spec_repos.py`
  - Removes need for bidict dependency
  - Allows both ssh and https urls to be used to determine course
  - Creates a list of supported courses
- Move `get_students` to separate file
- Use GitHub personal access token to authenticate gist instead of password
  - Using a password will be deprecated in November 2020 according to https://docs.github.com/en/rest/overview/other-authentication-methods#via-username-and-password

### CODE REVIEW

File: stograde/toolkit/args.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ def build_argparser():
 
 def get_students(args: Dict[str, Any]) -> List[str]:
     """Get students from the command line or the students.txt file.
-    Anything on the command line will override using the file"""
+    Any students on the command line will override using the file"""
```
Comment (hawkrives) [inline stograde/toolkit/args.py:145]:
```suggestion
    Passing any students on the command line will ignore the file contents"""
```

File: stograde/toolkit/args.py [hunk 2]
```diff
@@ -230,12 +222,11 @@ def process_args() -> Tuple[Dict[str, Any], List[str], List[str]]:
...
 
     else:
-        print('Sub-command must be specified')
+        print('Sub-command must be specified', file=sys.stderr)
         sys.exit(1)
 
     if not students:
-        print('No students selected\n'
-              'Is your students.txt missing?')
+        print('No students selected\nIs your students.txt missing?', file=sys.stderr)
```
Comment (hawkrives) [inline stograde/toolkit/args.py:229]:
```suggestion
        print('No students selected', file=sys.stderr)
        print('Is your students.txt missing?', file=sys.stderr)
```

File: stograde/toolkit/subcommands.py [hunk 3]
```diff
@@ -43,9 +43,15 @@ def do_ci(specs: Dict[str, 'Spec'],
                                                       workers=1,
                                                       work_dir='.')
 
-    passing: bool = ci_analyze(results)
+    # There should be only one student because the student name is
+    # retrieved from the environment by process_args,
+    # thus we can make the assumption that there is only one result
+    # and that it is at index 0
+    passing: bool = ci_analyze(results[0])
```
Comment (hawkrives) [inline stograde/toolkit/subcommands.py:50]: let's go ahead and add an `assert` here as well?

```suggestion
    assert len(results) == 1
    passing: bool = ci_analyze(results[0])
```

File: stograde/toolkit/args.py (hunk 1 above)
Comment (maxnz) [reply stograde/toolkit/args.py:145]: I like that much better

**Comment by maxnz** [comment]
Would we be able to have this reviewed and merged soon with a new release soon after? The bug fix listed at the beginning of the description is kinda important

**Pull request merged**